msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/StringHelper.java

```diff
@@ -42,6 +45,96 @@ static private String createSha256Hash(String stringToHash, boolean base64Encode
-        return str == null || str.trim().length() == 0;
+        return str == null || str.trim().isEmpty();
     }
-}
+
+    //Converts a map of parameters into a URL query string
+    static String serializeQueryParameters(Map<String, String> params) {
+        if (params != null && !params.isEmpty()) {
+            Map<String, String> encodedParams = urlEncodeMap(params);
+            StringBuilder sb = new StringBuilder();
+
+            for (Map.Entry<String, String> entry : encodedParams.entrySet()) {
+                if (entry.getKey() == null || entry.getValue() == null) {
+                    continue;
+                }
+
+                String value = entry.getValue();
+
+                if (sb.length() > 0) {
+                    sb.append('&');
+                }
+
+                sb.append(entry.getKey());
+                sb.append('=');
+                sb.append(value);
+            }
+
+            return sb.toString();
+        } else {
+            return "";
+        }
+    }
+
+    static Map<String, String> urlEncodeMap(Map<String, String> params) {
+        if (params == null || params.isEmpty()) {
+            return params;
+        } else {
+            Map<String, String> out = new LinkedHashMap<>();
+
+            for (Map.Entry<String, String> entry : params.entrySet()) {
+                try {
+                    String newKey = entry.getKey() != null ? URLEncoder.encode(entry.getKey(), "utf-8") : null;
+                    String newValue = entry.getValue() != null ? URLEncoder.encode(entry.getValue(), "utf-8") : null;
+
+                    out.put(newKey, newValue);
+                } catch (UnsupportedEncodingException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            return out;
+        }
+    }
+
+    static Map<String, String> parseQueryParameters(String query) {
+        Map<String, String> params = new LinkedHashMap<>();
+        if (StringHelper.isBlank(query)) {
+            return params;
+        } else {
+            StringTokenizer st = new StringTokenizer(query.trim(), "&");
+
+            while (st.hasMoreTokens()) {
+                String param = st.nextToken();
+                String[] pair = param.split("=", 2);
+
+                String key;
+                String value;
+                try {
+                    key = URLDecoder.decode(pair[0], "utf-8");
+                    value = pair.length > 1 ? URLDecoder.decode(pair[1], "utf-8") : "";
+                } catch (UnsupportedEncodingException | IllegalArgumentException e) {
+                    continue;
+                }
+
+                params.put(key, value);
+            }
+
+            return params;
+        }
+    }
+
+    //Much of the library once used Map<String, List<String>> for query parameters due to reliance on a specific dependency.
+    // Although Map<String, String> is now used internally, some public APIs still return Map<String, List<String>>
```

Comment:
This is not needed for all the cases? When you say some public APIs which ones?

Reply:
Just `AuthorizationRequestUrlParameters` returned query parameters as a `Map<String, List<String>>` in a public API, and in the latest commit I added that info the comment.